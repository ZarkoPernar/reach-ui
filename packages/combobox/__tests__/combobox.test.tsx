import React, { useState } from "react";
import { render, act, withMarkup } from "$test/utils";
import userEvent from "@testing-library/user-event";
import {
  Combobox,
  ComboboxInput,
  ComboboxList,
  ComboboxOption,
  ComboboxPopover,
} from "@reach/combobox";
import matchSorter from "match-sorter";
import cities from "../examples/cities";

describe("<Combobox />", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it("should match the snapshot", () => {
    let { baseElement, getByTestId, getByRole } = render(<BasicCombobox />);
    let input = getByTestId("input");
    expect(baseElement).toMatchSnapshot("before text input");
    userEvent.type(input, "e");
    expect(baseElement).toMatchSnapshot("after text input");
    expect(getByRole("combobox")).toHaveAttribute("aria-expanded", "true");
  });

  it("should open a list on text entry", () => {
    let optionToSelect = "Eagle Pass, Texas";

    let { getByTestId, getByText } = render(<BasicCombobox />);
    let getByTextWithMarkup = withMarkup(getByText);
    let input = getByTestId("input");
    userEvent.type(input, "e");
    act(() => void jest.advanceTimersByTime(100));
    expect(getByTestId("list")).toBeInTheDocument();
    expect(getByTextWithMarkup(optionToSelect)).toBeInTheDocument();
  });
});

////////////////////////////////////////////////////////////////////////////////
function BasicCombobox() {
  let [term, setTerm] = useState("");
  let results = useCityMatch(term);

  const handleChange = (event: any) => {
    setTerm(event.target.value);
  };

  return (
    <div>
      <h2>Clientside Search</h2>
      <Combobox id="holy-smokes">
        <ComboboxInput
          data-testid="input"
          name="awesome"
          onChange={handleChange}
        />
        {results ? (
          <ComboboxPopover portal={false}>
            <p>
              <button>Test focus</button>
            </p>
            <ComboboxList data-testid="list">
              {results.slice(0, 10).map((result, index) => (
                <ComboboxOption
                  key={index}
                  value={`${result.city}, ${result.state}`}
                />
              ))}
            </ComboboxList>
          </ComboboxPopover>
        ) : (
          <span>No Results!</span>
        )}
      </Combobox>
    </div>
  );
}

function useCityMatch(term: string) {
  return term.trim() === ""
    ? null
    : matchSorter(cities, term, {
        keys: [item => `${item.city}, ${item.state}`],
      });
}
