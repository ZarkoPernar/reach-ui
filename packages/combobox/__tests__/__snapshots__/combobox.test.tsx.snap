// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Combobox /> should match the snapshot: after text input 1`] = `
<body>
  <div>
    <div>
      <h2>
        Clientside Search
      </h2>
      <div
        aria-expanded="true"
        aria-haspopup="listbox"
        aria-owns="listbox--holy-smokes"
        data-reach-combobox=""
        id="holy-smokes"
        role="combobox"
      >
        <input
          aria-autocomplete="both"
          aria-controls="listbox--holy-smokes"
          data-reach-combobox-input=""
          data-testid="input"
          name="awesome"
          value="e"
        />
        <div
          data-reach-combobox-popover=""
          tabindex="-1"
        >
          <p>
            <button>
              Test focus
            </button>
          </p>
          <ul
            data-reach-combobox-list=""
            data-testid="list"
            id="listbox--holy-smokes"
            role="listbox"
          >
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="931144230"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                agan, Minn
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                sota
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-276154372"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                agl
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                 Mountain, T
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                xas
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="1710908408"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                agl
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                 Pass, T
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                xas
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-1574158667"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                agl
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                , Idaho
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-937984427"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                arlimart, California
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-203649349"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                asl
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                y, South Carolina
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="447155012"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                ast Alton, Illinois
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-1162835674"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                ast Aurora, N
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                w York
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="1182450877"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                ast B
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                th
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                l, Minn
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                sota
              </span>
            </li>
            <li
              aria-selected="false"
              data-reach-combobox-option=""
              id="-190973284"
              role="option"
              tabindex="-1"
            >
              <span
                data-user-value="true"
              >
                E
              </span>
              <span
                data-suggested-value="true"
              >
                ast Brain
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                rd, T
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                nn
              </span>
              <span
                data-user-value="true"
              >
                e
              </span>
              <span
                data-suggested-value="true"
              >
                ss
              </span>
              <span
                data-user-value="true"
              >
                ee
              </span>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`<Combobox /> should match the snapshot: before text input 1`] = `
<body>
  <div>
    <div>
      <h2>
        Clientside Search
      </h2>
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="listbox--holy-smokes"
        data-reach-combobox=""
        id="holy-smokes"
        role="combobox"
      >
        <input
          aria-autocomplete="both"
          aria-controls="listbox--holy-smokes"
          data-reach-combobox-input=""
          data-testid="input"
          name="awesome"
          value=""
        />
        <span>
          No Results!
        </span>
      </div>
    </div>
  </div>
</body>
`;
